src/sx/src/injectRuntimeStyles.js

```diff
@@ -73,42 +44,41 @@ export function injectRuntimeKeyframes(css: string, name: string) {
 }
 
 // https://developer.mozilla.org/en-US/docs/Web/API/CSS_Object_Model
-export default function injectRuntimeStyles(styleBuffer: StyleBufferType) {
-  const styleSheet = getStyleTag();
-
-  const matchFunction = (node) => (cssRule) => {
-    if (cssRule.type === CSSRule.STYLE_RULE) {
-      const styleRule = ((cssRule: any): CSSStyleRule);
-      if (styleRule.selectorText === `.${node.getHash()}`) {
-        return true;
-      }
-    }
-    return false;
-  };
+export default function injectRuntimeStyles(
+  styleSheet: CSSStyleSheet, // TODO: note: we are mutating this
+  { rehydratedStyleRules, rehydratedMediaRules, rehydratedKeyframeRules }: RehydratedStyles,
+  styleBuffer: StyleBufferType,
+) {
+  // const { rehydratedStyleRules, rehydratedMediaRules, rehydratedKeyframeRules } = x;
 
   styleBuffer.forEach((node) => {
     const insertIndex = styleSheet.cssRules.length;
 
     if (node instanceof StyleCollectorNode) {
-      if (hasStyleRule(matchFunction(node)) === false) {
+      if (rehydratedStyleRules.has(`.${node.getHash()}`) === false) {
         // apply missing styles
         const rule = serialize(
           compile(node.printNodes({ trailingSemicolon: true }).join('')),
           middleware([prefixer, stringify]),
         );
+
+        // Apply the missing styles:
         styleSheet.insertRule(rule, insertIndex);
+
+        // Register the rule into rehydrated styles, so we don't attempt to process it again:
+        rehydratedStyleRules.add(`.${node.getHash()}`);
       }
     } else if (node instanceof StyleCollectorAtNode) {
-      // TODO: make sure we are not adding already added styles (?)
+      // TODO: leverage rehydrated rules (to add only new rules)
```

Comment:
There is still one missing piece -> it's difficult to convert between `StyleCollectorAtNode` and the rehydrated style identifier. I have this change WIP locally as well.